File: backend/app/sandbox/docker_backend.py
"""Docker-based sandbox backend.

Each execution runs in a throwaway, heavily-restricted container:
  - no network (network_mode="none")
  - capped memory (+ no swap), CPU, and PIDs
  - read-only root filesystem with a small tmpfs at /sandbox for the code file
  - all Linux capabilities dropped + no-new-privileges
  - non-root user (baked into the image)
  - wall-clock timeout enforced here (the container is killed if it overruns)
  - auto-removed so nothing persists between runs

The code is passed via stdin to a tiny runner so we never mount host paths or rely on the
container filesystem being writable beyond the tmpfs.
"""
import time

import docker
from docker.errors import ContainerError, ImageNotFound

from app.sandbox.base import ExecutionLimits, SandboxBackend
from app.schemas import SandboxResult

# Map of supported languages -> how to run a single source file inside the sandbox image.
# Extending to new languages is a matter of adding an entry here (and to the image).
_LANG_RUNNERS = {
    "python": {
        "filename": "code.py",
        "cmd": ["python", "-I", "-B", "/sandbox/code.py"],
    },
}

_WORKDIR = "/sandbox"


class DockerBackend(SandboxBackend):
    def __init__(self, image: str):
        self._image = image
        self._client = docker.from_env()

    def execute(self, code: str, language: str, limits: ExecutionLimits) -> SandboxResult:
        runner = _LANG_RUNNERS.get(language)
        if runner is None:
            return SandboxResult(
                stdout="",
                stderr=f"Unsupported language: {language!r}",
                exit_code=2,
                duration_ms=0,
                timed_out=False,
            )

        # The container writes the provided code to the tmpfs and executes it. Using a
        # shell heredoc avoids mounting any host path into the untrusted container.
        bootstrap = (
            f"cat > {_WORKDIR}/{runner['filename']} <<'__LLM_EOF__'\n"
            f"{code}\n"
            "__LLM_EOF__\n"
            "exec " + " ".join(runner["cmd"])
        )

        container = None
        started = time.monotonic()
        timed_out = False
        try:
            container = self._client.containers.run(
                self._image,
                command=["sh", "-c", bootstrap],
                detach=True,
                # --- isolation / hardening ---
                network_mode="none",
                mem_limit=f"{limits.memory_mb}m",
                memswap_limit=f"{limits.memory_mb}m",  # == mem_limit disables swap
                nano_cpus=int(limits.cpus * 1_000_000_000),
                pids_limit=limits.pids_limit,
                read_only=True,
                # Writable, ephemeral scratch space: /sandbox holds the code file,
                # /tmp is general scratch (the system prompt promises code a writable /tmp).
                # Everything else stays read-only.
                tmpfs={_WORKDIR: "rw,size=8m,mode=1777", "/tmp": "rw,size=16m,mode=1777"},
                cap_drop=["ALL"],
                security_opt=["no-new-privileges"],
                user="1000:1000",
                working_dir=_WORKDIR,
                # Keep stdout/stderr separate so we can report them distinctly.
                stdout=True,
                stderr=True,
            )

            try:
                result = container.wait(timeout=limits.timeout_seconds)
                exit_code = int(result.get("StatusCode", -1))
            except Exception:
                # Timed out (or daemon hiccup): force-kill and report a timeout.
                timed_out = True
                exit_code = 124
                try:
                    container.kill()
                except Exception:
                    pass

            duration_ms = int((time.monotonic() - started) * 1000)

            stdout = container.logs(stdout=True, stderr=False).decode("utf-8", "replace")
            stderr = container.logs(stdout=False, stderr=True).decode("utf-8", "replace")
            if timed_out:
                stderr = (
                    stderr + f"\n[sandbox] killed after {limits.timeout_seconds}s timeout"
                ).strip()

            return SandboxResult(
                stdout=_truncate(stdout, limits.max_output_chars),
                stderr=_truncate(stderr, limits.max_output_chars),
                exit_code=exit_code,
                duration_ms=duration_ms,
                timed_out=timed_out,
            )
        except ImageNotFound:
            return SandboxResult(
                stdout="",
                stderr=(
                    f"[sandbox] image {self._image!r} not found. Build it first: "
                    "`docker build -t llm-sandbox:latest backend/sandbox-image`."
                ),
                exit_code=1,
                duration_ms=int((time.monotonic() - started) * 1000),
                timed_out=False,
            )
        except ContainerError as exc:  # non-zero exit with combined output
            return SandboxResult(
                stdout="",
                stderr=_truncate(str(exc), limits.max_output_chars),
                exit_code=exc.exit_status,
                duration_ms=int((time.monotonic() - started) * 1000),
                timed_out=False,
            )
        finally:
            if container is not None:
                try:
                    container.remove(force=True)  # ensure --rm semantics even on timeout
                except Exception:
                    pass


def _truncate(text: str, limit: int) -> str:
    if len(text) <= limit:
        return text
    return text[:limit] + f"\n…[truncated, {len(text) - limit} more chars]"

File: backend/src/sandbox/dockerBackend.ts
/**
 * Docker-based sandbox backend (port of docker_backend.py). Each execution runs in a
 * throwaway, heavily-restricted container:
 *   - no network (NetworkMode "none")
 *   - capped memory (+ no swap), CPU, and PIDs
 *   - read-only root filesystem with small tmpfs at /sandbox (code file) and /tmp
 *   - all Linux capabilities dropped + no-new-privileges
 *   - non-root user (baked into the image; also enforced here)
 *   - wall-clock timeout enforced here (the container is killed if it overruns)
 *   - force-removed so nothing persists between runs (--rm semantics)
 *
 * The code is passed via stdin to a tiny shell runner so we never mount host paths or
 * rely on the container filesystem being writable beyond the tmpfs.
 */
import Docker from "dockerode";
import { PassThrough } from "node:stream";
import type { SandboxBackend, ExecutionLimits } from "./base.js";
import type { SandboxResult } from "../schemas.js";

const LANG_RUNNERS: Record<string, { filename: string; cmd: string[] }> = {
  python: { filename: "code.py", cmd: ["python", "-I", "-B", "/sandbox/code.py"] },
};
const WORKDIR = "/sandbox";

const delay = (ms: number) => new Promise((r) => setTimeout(r, ms));
const elapsedMs = (start: bigint) => Number((process.hrtime.bigint() - start) / 1_000_000n);

function truncate(text: string, limit: number): string {
  if (text.length <= limit) return text;
  return `${text.slice(0, limit)}\n…[truncated, ${text.length - limit} more chars]`;
}

export class DockerBackend implements SandboxBackend {
  private readonly docker: Docker;

  constructor(private readonly image: string) {
    this.docker = new Docker(); // defaults to /var/run/docker.sock
  }

  async execute(code: string, language: string, limits: ExecutionLimits): Promise<SandboxResult> {
    const runner = LANG_RUNNERS[language];
    if (!runner) {
      return {
        stdout: "",
        stderr: `Unsupported language: '${language}'`,
        exitCode: 2,
        durationMs: 0,
        timedOut: false,
      };
    }

    // The container writes the provided code to the tmpfs and executes it. A shell
    // heredoc avoids mounting any host path into the untrusted container.
    const bootstrap =
      `cat > ${WORKDIR}/${runner.filename} <<'__LLM_EOF__'\n` +
      `${code}\n` +
      `__LLM_EOF__\n` +
      `exec ${runner.cmd.join(" ")}`;

    const started = process.hrtime.bigint();
    let container: Docker.Container | undefined;
    let timedOut = false;

    try {
      container = await this.docker.createContainer({
        Image: this.image,
        Cmd: ["sh", "-c", bootstrap],
        Tty: false,
        WorkingDir: WORKDIR,
        User: "1000:1000",
        HostConfig: {
          NetworkMode: "none",
          Memory: limits.memoryMb * 1024 * 1024,
          MemorySwap: limits.memoryMb * 1024 * 1024, // == Memory disables swap
          NanoCpus: Math.round(limits.cpus * 1_000_000_000),
          PidsLimit: limits.pidsLimit,
          ReadonlyRootfs: true,
          Tmpfs: {
            [WORKDIR]: "rw,size=8m,mode=1777",
            "/tmp": "rw,size=16m,mode=1777",
          },
          CapDrop: ["ALL"],
          SecurityOpt: ["no-new-privileges"],
          // Note: AutoRemove intentionally NOT set — we force-remove in finally so we
          // can reliably read logs and enforce the timeout kill first (== Python --rm).
        },
      });

      // Attach BEFORE start so no early output is lost; demux into two buffers.
      const stdoutChunks: Buffer[] = [];
      const stderrChunks: Buffer[] = [];
      const outStream = new PassThrough();
      const errStream = new PassThrough();
      outStream.on("data", (c: Buffer) => stdoutChunks.push(c));
      errStream.on("data", (c: Buffer) => stderrChunks.push(c));
      const attach = await container.attach({ stream: true, stdout: true, stderr: true });
      this.docker.modem.demuxStream(attach, outStream, errStream);
      const streamEnded = new Promise<void>((resolve) => {
        attach.on("end", () => resolve());
        attach.on("close", () => resolve());
        attach.on("error", () => resolve());
      });

      await container.start();

      const waitPromise = container.wait();
      const timeoutMs = limits.timeoutSeconds * 1000;
      // Cancelable timer: clear it when wait wins so we don't leave a live ref'd
      // setTimeout per execution (they'd accumulate for the full timeout under load).
      let timeoutHandle: ReturnType<typeof setTimeout> | undefined;
      const timeoutPromise = new Promise<{ kind: "timeout" }>((resolve) => {
        timeoutHandle = setTimeout(() => resolve({ kind: "timeout" as const }), timeoutMs);
      });
      let outcome: { kind: "done"; code: number } | { kind: "timeout" };
      try {
        outcome = await Promise.race([
          waitPromise.then((r) => ({ kind: "done" as const, code: Number(r?.StatusCode ?? -1) })),
          timeoutPromise,
        ]);
      } finally {
        if (timeoutHandle) clearTimeout(timeoutHandle);
      }

      let exitCode: number;
      if (outcome.kind === "timeout") {
        timedOut = true;
        exitCode = 124;
        await container.kill().catch(() => {});
        await waitPromise.catch(() => {}); // let wait settle after the kill
      } else {
        exitCode = outcome.code;
      }

      // Bounded flush of the attach stream so buffered output lands.
      await Promise.race([streamEnded, delay(500)]);

      const durationMs = elapsedMs(started);
      const stdout = Buffer.concat(stdoutChunks).toString("utf8");
      let stderr = Buffer.concat(stderrChunks).toString("utf8");
      if (timedOut) {
        stderr = `${stderr}\n[sandbox] killed after ${limits.timeoutSeconds}s timeout`.trim();
      }

      return {
        stdout: truncate(stdout, limits.maxOutputChars),
        stderr: truncate(stderr, limits.maxOutputChars),
        exitCode,
        durationMs,
        timedOut,
      };
    } catch (err) {
      const e = err as { statusCode?: number; message?: string };
      if (e?.statusCode === 404) {
        return {
          stdout: "",
          stderr:
            `[sandbox] image '${this.image}' not found. Build it first: ` +
            "`docker build -t llm-sandbox:latest backend/sandbox-image`.",
          exitCode: 1,
          durationMs: elapsedMs(started),
          timedOut: false,
        };
      }
      return {
        stdout: "",
        stderr: truncate(String(e?.message ?? err), limits.maxOutputChars),
        exitCode: 1,
        durationMs: elapsedMs(started),
        timedOut: false,
      };
    } finally {
      if (container) {
        await container.remove({ force: true }).catch(() => {}); // ensure --rm even on timeout
      }
    }
  }
}
